File: scripts/jpx/update_stock_master.py
#!/usr/bin/env python3
"""
Stock マスタデータベース更新スクリプト

JPXからスクレイピングした銘柄データ（jpx_stocks.json）をPostgreSQLに反映する。

実行方法:
  DATABASE_URL="postgresql://..." python scripts/jpx/update_stock_master.py

前提条件:
  - scripts/jpx/jpx_stocks.json が存在すること
  - DATABASE_URL が設定されていること
"""

import json
import os
import re
import sys
from pathlib import Path

import psycopg2
import psycopg2.extras

# バッチサイズ
BATCH_SIZE = 100


def get_database_url() -> str:
    """データベースURLを取得"""
    url = os.environ.get("PRODUCTION_DATABASE_URL") or os.environ.get("DATABASE_URL")
    if not url:
        print("Error: DATABASE_URL environment variable not set")
        sys.exit(1)
    return url


def load_json_data(file_path: str) -> list[dict]:
    """JSONファイルからデータを読み込む"""
    try:
        with open(file_path, "r", encoding="utf-8") as f:
            stocks = json.load(f)
        print(f"Loaded {len(stocks)} records from {file_path}")
        return stocks
    except FileNotFoundError:
        print(f"File not found: {file_path}")
        print("Run scrape_stocks.py first to generate the data file.")
        sys.exit(1)
    except Exception as e:
        print(f"Error loading JSON: {e}")
        sys.exit(1)


def update_stock_master(conn, stocks: list[dict]) -> dict:
    """銘柄マスタを更新"""
    if not stocks:
        print("No stocks to update")
        return {"added": 0, "updated": 0, "errors": 0}

    print("Processing stocks...")

    added = 0
    updated = 0
    errors = 0

    # 重複除去とバリデーション
    seen_tickers: set[str] = set()
    unique_stocks: list[dict] = []

    date_regex = re.compile(r"^\d{4}-\d{2}-\d{2}$")

    for stock in stocks:
        ticker = stock.get("ticker")
        name = stock.get("name")

        if not ticker or not name:
            print(f"  Skipping invalid record: {stock}")
            errors += 1
            continue

        if ticker in seen_tickers:
            print(f"  Skipping duplicate ticker: {ticker}")
            continue
        seen_tickers.add(ticker)

        # listedDate の検証
        listed_date = stock.get("listedDate")
        if listed_date and not date_regex.match(listed_date):
            print(f"  Invalid date format for {ticker}: {listed_date}")
            stock["listedDate"] = None

        unique_stocks.append(stock)

    print(f"Upserting {len(unique_stocks)} stocks to database...")

    with conn.cursor() as cur:
        for i in range(0, len(unique_stocks), BATCH_SIZE):
            batch = unique_stocks[i : i + BATCH_SIZE]

            try:
                tickers = [s["ticker"] for s in batch]

                # 既存のtickerCodeを確認
                cur.execute(
                    'SELECT "tickerCode" FROM "Stock" WHERE "tickerCode" = ANY(%s)',
                    (tickers,),
                )
                existing_tickers = {row[0] for row in cur.fetchall()}

                to_create = [s for s in batch if s["ticker"] not in existing_tickers]
                to_update = [s for s in batch if s["ticker"] in existing_tickers]

                # 新規追加
                if to_create:
                    psycopg2.extras.execute_values(
                        cur,
                        """
                        INSERT INTO "Stock" ("tickerCode", name, market, sector, "createdAt")
                        VALUES %s
                        ON CONFLICT ("tickerCode") DO NOTHING
                        """,
                        [
                            (
                                s["ticker"],
                                s["name"],
                                "TSE",
                                s.get("sector") or "その他",
                            )
                            for s in to_create
                        ],
                        template="(%s, %s, %s, %s, NOW())",
                        page_size=BATCH_SIZE,
                    )
                    added += len(to_create)

                # 更新
                for stock in to_update:
                    cur.execute(
                        """
                        UPDATE "Stock"
                        SET name = %s, sector = COALESCE(%s, sector)
                        WHERE "tickerCode" = %s
                        """,
                        (stock["name"], stock.get("sector"), stock["ticker"]),
                    )
                    updated += 1

                batch_num = i // BATCH_SIZE + 1
                print(f"  Batch {batch_num}: {len(to_create)} added, {len(to_update)} updated")

            except Exception as e:
                batch_num = i // BATCH_SIZE + 1
                print(f"  Error in batch {batch_num}: {e}")
                errors += len(batch)

    conn.commit()

    print()
    print("=" * 60)
    print("Database update completed:")
    print(f"  Added: {added}")
    print(f"  Updated: {updated}")
    print(f"  Errors: {errors}")
    print("=" * 60)

    return {"added": added, "updated": updated, "errors": errors}


def main() -> int:
    print("=" * 60)
    print("Stock Master Update Script")
    print("=" * 60)
    print()

    # JSONファイルのパス
    script_dir = Path(__file__).parent
    json_file = script_dir / "jpx_stocks.json"

    # JSONデータを読み込み
    stocks = load_json_data(str(json_file))
    print()

    if not stocks:
        print("No stocks to process. Exiting.")
        return 0

    # DB接続
    db_url = get_database_url()
    conn = psycopg2.connect(db_url)

    try:
        # データベースを更新
        stats = update_stock_master(conn, stocks)

        # 終了コードを決定
        if stats["errors"] > len(stocks) * 0.5:
            print(f"\nToo many errors ({stats['errors']}/{len(stocks)})")
            return 1

        print("\nUpdate completed successfully!")
        return 0

    except Exception as e:
        print(f"Error: {e}")
        return 1
    finally:
        conn.close()


if __name__ == "__main__":
    sys.exit(main())

File: scripts/jpx/update-stock-master.ts
#!/usr/bin/env npx tsx
/**
 * Stock マスタデータベース更新スクリプト
 *
 * JPXからスクレイピングした銘柄データをPostgreSQLに反映する。
 *
 * 実行方法:
 *   export PRODUCTION_DATABASE_URL="postgresql://..."
 *   npx tsx scripts/jpx/update-stock-master.ts
 *
 * 前提条件:
 *   - scripts/jpx/jpx_stocks.json が存在すること
 *   - PRODUCTION_DATABASE_URL が設定されていること
 */

import { PrismaClient } from "@prisma/client"
import * as fs from "fs"
import * as path from "path"

// PRODUCTION_DATABASE_URL を使用（設定されていない場合はDATABASE_URLを使用）
const databaseUrl = process.env.PRODUCTION_DATABASE_URL || process.env.DATABASE_URL

if (!databaseUrl) {
  console.error("ERROR: PRODUCTION_DATABASE_URL or DATABASE_URL environment variable is not set")
  process.exit(1)
}

const prisma = new PrismaClient({
  datasources: {
    db: {
      url: databaseUrl,
    },
  },
})

interface StockData {
  ticker: string
  name: string
  sector?: string | null
  listedDate?: string | null
}

function loadJsonData(filePath: string): StockData[] {
  try {
    const data = fs.readFileSync(filePath, "utf-8")
    const stocks = JSON.parse(data) as StockData[]
    console.log(`Loaded ${stocks.length} records from ${filePath}`)
    return stocks
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") {
      console.error(`File not found: ${filePath}`)
      console.log("Run scrape-stocks.ts first to generate the data file.")
    } else {
      console.error(`Error loading JSON: ${error}`)
    }
    process.exit(1)
  }
}

async function updateStockMaster(stocks: StockData[]): Promise<{ added: number; updated: number; errors: number }> {
  if (stocks.length === 0) {
    console.log("No stocks to update")
    return { added: 0, updated: 0, errors: 0 }
  }

  console.log("Connecting to database...")

  let added = 0
  let updated = 0
  let errors = 0

  // 重複除去
  const seenTickers = new Set<string>()
  const uniqueStocks: StockData[] = []

  for (const stock of stocks) {
    if (!stock.ticker || !stock.name) {
      console.log(`Skipping invalid record: ${JSON.stringify(stock)}`)
      errors++
      continue
    }

    if (seenTickers.has(stock.ticker)) {
      console.log(`Skipping duplicate ticker: ${stock.ticker}`)
      continue
    }
    seenTickers.add(stock.ticker)

    // listedDate の検証
    if (stock.listedDate) {
      const dateRegex = /^\d{4}-\d{2}-\d{2}$/
      if (!dateRegex.test(stock.listedDate)) {
        console.log(`Invalid date format for ${stock.ticker}: ${stock.listedDate}`)
        stock.listedDate = null
      }
    }

    uniqueStocks.push(stock)
  }

  console.log(`Upserting ${uniqueStocks.length} stocks to database...`)

  // バッチ処理（100件ずつ）
  const batchSize = 100

  for (let i = 0; i < uniqueStocks.length; i += batchSize) {
    const batch = uniqueStocks.slice(i, i + batchSize)

    try {
      // 既存のtickerCodeを確認
      const tickers = batch.map((s) => s.ticker)
      const existingStocks = await prisma.stock.findMany({
        where: { tickerCode: { in: tickers } },
        select: { tickerCode: true },
      })
      const existingTickers = new Set(existingStocks.map((s) => s.tickerCode))

      // 新規追加と更新を分離
      const toCreate = batch.filter((s) => !existingTickers.has(s.ticker))
      const toUpdate = batch.filter((s) => existingTickers.has(s.ticker))

      // 新規追加
      if (toCreate.length > 0) {
        await prisma.stock.createMany({
          data: toCreate.map((s) => ({
            tickerCode: s.ticker,
            name: s.name,
            market: "東証プライム",
            sector: s.sector || "その他",
            listedDate: s.listedDate ? new Date(s.listedDate) : null,
          })),
          skipDuplicates: true,
        })
        added += toCreate.length
      }

      // 更新
      for (const stock of toUpdate) {
        await prisma.stock.update({
          where: { tickerCode: stock.ticker },
          data: {
            name: stock.name,
            sector: stock.sector || undefined,
            listedDate: stock.listedDate ? new Date(stock.listedDate) : undefined,
          },
        })
        updated++
      }

      console.log(`  Batch ${Math.floor(i / batchSize) + 1}: ${toCreate.length} added, ${toUpdate.length} updated`)
    } catch (error) {
      console.log(`Error in batch ${Math.floor(i / batchSize) + 1}: ${error}`)
      errors += batch.length
    }
  }

  console.log()
  console.log("=".repeat(60))
  console.log("Database update completed:")
  console.log(`  Added: ${added}`)
  console.log(`  Updated: ${updated}`)
  console.log(`  Errors: ${errors}`)
  console.log("=".repeat(60))

  return { added, updated, errors }
}

async function main(): Promise<void> {
  console.log("=".repeat(60))
  console.log("Stock Master Update Script")
  console.log("=".repeat(60))
  console.log()

  // JSONファイルのパス
  const jsonFile = path.join(__dirname, "jpx_stocks.json")

  // JSONデータを読み込み
  const stocks = loadJsonData(jsonFile)
  console.log()

  if (stocks.length === 0) {
    console.log("No stocks to process. Exiting.")
    return
  }

  try {
    // データベースを更新
    const stats = await updateStockMaster(stocks)

    // 終了コードを決定
    if (stats.errors > stocks.length * 0.5) {
      console.log(`\nToo many errors (${stats.errors}/${stocks.length})`)
      process.exit(1)
    }

    console.log("\nUpdate completed successfully!")
  } catch (error) {
    console.error(`Error: ${error}`)
    process.exit(1)
  } finally {
    await prisma.$disconnect()
  }
}

main()

export {}
